Add table block caption

talldan opened the pull request:
Related #15283, #6923

Fixes #11589

## Description
- Adds the ability to specify a caption for the table block.

## How has this been tested?
1. Add a table
2. Add a caption to the table
3. Preview the post and check that the table has a caption

## Screenshots <!-- if applicable -->
Side-by-side of image and table captions in the editor
![Screen Shot 2019-05-10 at 4 00 23 pm](https://user-images.githubusercontent.com/677833/57512082-4dfaeb80-733d-11e9-8ceb-a63b36d17161.png)


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

talldan committed:
Initial attempt at table block caption

Try using a figcaption for the table caption

Revert "Try using a figcaption for the table caption"

Switch back to a standard table caption

This reverts commit b5f00ddfb3bba0f4f2356bfe83e78e75932ded48.

Add comment explaining location of RichText

Ensure table cell is deselected when navigating from a table cell to the caption

Styling tweaks

Add an inline toolbar to the caption to match the image block implementation

Adjust margin to match image block

Add e2e test for the table block caption

Refine caption accessibility.

Use a div for the caption content

Update caption to use a figcaption element with aria-labelledby

Update table block deprecation so that it has its own attribute definition

Update existing fixtures and add a new fixture for table captions

Minor refinements

Fix hasCaption logic

Update snapshots

Update comment with latest understanding of issue

Use client id to generate non-clashing caption element id

Update block-transforms fixture

Update test to use a regular expression

talldan commented:
It'd be great to get some accessibility feedback on this PR, especially on the editor implementation of the caption.

Initially I thought this would be easy, but the tricky part is that the html table element has only a very specific set of elements it allows as children. I wasn't able to use a `<RichText>` directly within the table, as that component outputs a `<div>` as its root element.

Instead I've added a normal caption tag to the table and made it `visibility: hidden` (not sure if making it zero height and overflow hidden would be preferable?). Then I've added a RichText outside the table for editing the caption.

Is there a better way to do this?

talldan commented:
Think that adding a `rootTagName` prop to `RichText` might be a better solution so that I don't have to add a hack.

talldan commented:
Well ... I don't know why I didn't think of this before, but the obvious solution is to put the `RichText` component in the `caption` element. Not sure why I didn't think of this before.🤦‍♂  

However, I tested it out, and that introduces other problems. The HTML spec mentions to put the `caption` before the `tbody` in a `table`, but if the caption is first in the DOM the tab order is wrong for users of the table block.

The tbody also overlaps the caption's inline toolbar, and adjusting the z-index doesn't seem to resolve this. 🤔 

Perhaps the original solution was the better one.

aduth commented:
@talldan I think it could be appropriate here to use the [`figure` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figure)? In general, I'd see this as being implemented almost identical to the image block (though I'm not really a fan of its `save` implementation with always-`figure` and `div` wrappers). 

```html
<figure>
    <table><!-- ... --></table>
    <figcaption>My lovely caption</figcaption>
</figure>
```

See also: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table#attr-summary

>Include the `<table>` element in a `<figure>` element and add the description in prose next to it.

afercia commented:
Re: the div wrappers problem within the editor:
I'm not sure if it's technically possible, but I'd vote for a contenteditable `<caption>` element. Is it possible to use that instead of a RichText component?

Also, while HTML allows to [omit the caption element when the table is wrapped in a figure with a figcaption](https://www.w3.org/TR/html52/tabular-data.html#the-caption-element), I'm not sure about actual support from assistive technologies. I'd recommend to use the `<caption>` element.

Hiding the `<caption>` element seems a bit hacky to me, as it wouldn't be perceived by assistive technologies and as a consequence the table wouldn't have an associated title.

If it's not possible to use a `<caption>` within the editor, I guess we should just accept the table won't have a 100% correct semantic in the editor.

The default placement of the caption should be at the top. That's the default for table captions. The placement at the bottom could be implemented as a style variation. Worth noting a table caption is something different from what generally Gutenberg means for "captions". Therefore, I'd also propose to rename this field to "Table caption".

talldan commented:
Many thanks for the feedback, @aduth and @afercia. I haven't had much time lately to develop this further. I'll do some thinking on this, and try to come up with a way forward. It sounds like I also need to do some testing to understand the difference between how caption and figcaption might work in assistive tech.

talldan commented:
Have been doing some initial testing in VoiceOver of the various options for the table caption implementation. (cc: @afercia, would be interested in your thoughts on this). 

I've implemented option 2 described below on this branch if anyone wants to test it.

**Option 1 (RichText inside the caption element)**
```
<table>
    <caption><RichText ... /></caption>
     // ...
</table>
```
For:
- Correct table markup

Against:
- I haven't figured out a way to make the tab order work when using `caption-side: bottom;`, so this option might be limited to only displaying the caption above the table.
- VoiceOver didn't seem to know what to do with a contenteditable inside a table caption. It doesn't seem to announce it when editing. That would seem to rule this option out.

**Option 2 (RichText outside the table, visibly hidden caption)**
```
<table>
    <caption className="screen-reader-text">{ caption }</caption>
     // ...
</table>
<RichText ... />
```

Pros:
- Correct table markup.
- Could support both `caption-side: top;` and `caption-side: bottom;`.

Cons:
- The user has to leave the context of the table to edit the caption. Though they're at least still within the context of the table block.
- After altering the content of the RichText, the updated table caption wasn't read by VoiceOver in Chrome (worked fine in Safari). Forcing the element to be replaced in the DOM instead of updated resolved this, but it's a bit of a hack!

**Option 3 (figure/figcaption)**
```
<figure>
    <table>
         // ...
    </table>
     <figcaption><RichText /></figcaption>
<figure>
```

Pros:
- Could support both `caption-side: top;` and `caption-side: bottom;`.

Cons:
- While this is considered ok, it's not quite how table captions are supposed to be marked up.
- The caption isn't read by voiceover when interacting with the table.
- The user has to leave the context of the table to edit the caption. Though they're at least still within the context of the table block.

afercia commented:
@talldan thanks for working on this. Right, I tested also on Windows and seems most of the browser / screen reader combinations have troubles with the editable within the `<caption>` element. Some of them focus the editable but don't announce anything. Some don't even focus the editable.

Codepen for testing: https://codepen.io/afercia/full/RzbMvx

The same happens with a native `<input>` field so this appears to be some unfortunate browsers limitation.

Will ping around to see if others have faced the same issue. For now, I don't have suggestions other than using a div element with an `aria-label="Table caption"` to clarify what it is. The important thing is that in the front end it's rendered as a real `<caption>` element.

talldan commented:
Thanks @afercia. Hopefully what I've done so far is the best compromise. The contenteditable has the aria label 'Write caption ...' at the moment, which matches the image block.

talldan commented:
I've added the 'Needs Design Feedback' label.

On this branch I've added an option for toggling the caption position, which might need some finesse 😄 

At the moment I've added this in the toolbar menu that contains other table options (insert row, delete row etc.).

Questions:
- Is this the right place for this option, or should it be somewhere else? Maybe the sidebar? I considered that, but wasn't sure what kind of control would be right for the option. A dropdown?
- If a toolbar button is a good choice, it'll probably need a better icon (and maybe better copy). Any suggestions? I had a look for an icon but couldn't find anything suitable, it might need a custom design.

Here's how it looks at the moment:
![Screen Shot 2019-06-11 at 7 26 02 pm](https://user-images.githubusercontent.com/677833/59268654-c5df6d00-8c7f-11e9-867b-0220b63eae48.png)

Once the button is clicked, the text and the icon change:
![Screen Shot 2019-06-11 at 7 34 12 pm](https://user-images.githubusercontent.com/677833/59268711-ead3e000-8c7f-11e9-8d81-5ba22a6bc9fe.png)

Thanks! Also, any other general design feedback on this feature is appreciated.

kjellr commented:
> Is this the right place for this option, or should it be somewhere else? Maybe the sidebar? I considered that, but wasn't sure what kind of control would be right for the option. A dropdown?
> 
> If a toolbar button is a good choice, it'll probably need a better icon (and maybe better copy). Any suggestions? I had a look for an icon but couldn't find anything suitable, it might need a custom design.

Hmm yeah. I do think that option is in a weird place right now. Placing it into its own separate toolbar item makes sense. 

Regarding the icon, this is one of those cases where choosing one is way more complicated than it should be. Ideally, the two icons would look like this: 

<img width="29" alt="Screen Shot 2019-06-27 at 9 49 32 AM" src="https://user-images.githubusercontent.com/1202812/60272039-690bc580-98c1-11e9-89ff-e0f52667ed06.png">

<img width="29" alt="Screen Shot 2019-06-27 at 9 49 32 AM" src="https://user-images.githubusercontent.com/1202812/60273885-13391c80-98c5-11e9-8aa4-36d29a620b3e.png">

... but unfortunately one of those icons is already our icon for "Full width". 😄  

Since that's the case, perhaps something like this would work for us: 

<img width="258" alt="Screen Shot 2019-06-27 at 10 25 19 AM" src="https://user-images.githubusercontent.com/1202812/60274365-eb968400-98c5-11e9-90ae-3b009f3ad920.png">

[Here are SVGs for those](https://cloudup.com/cG-zTKsNbHT) — I'd love some other 👀 on them too though. Perhaps @melchoyce if you have a moment? I think this icon + functionality could end up being reused for other blocks in the future.

talldan commented:
Thanks for all the feedback, @kjellr 😄 

> Hmm yeah. I do think that option is in a weird place right now. Placing it into its own separate toolbar item makes sense.

I'll work on that. I imagine it'll require collapsing some of the existing toolbar options for the table block. There's a separate PR adding cell alignment options to the toolbar, and the toolbar is looking quite wide.

Not sure if there are any guides on how many icons on the toolbar is too many.

kjellr commented:
There's a little bit of thought on that happening in https://github.com/WordPress/gutenberg/issues/15096. (cc @jasmussen). My general thought is that 3 toolbar icons for each of the "sections" outlined there makes sense.

talldan commented:
Testing it out some more, and I feel like an option might be to only show the caption position button when the caption has selection. At the same time the normal table menu could be hidden when the caption has selection (the options within can only be used when a cell is selected).

kjellr commented:
> Testing it out some more, and I feel like an option might be to only show the caption position button when the caption has selection. At the same time the normal table menu could be hidden when the caption has selection (the options within can only be used when a cell is selected).

That's a good point. Also, it may make sense to adopt the floating inline controls for captions that we use for other block captions: 

![Screen Shot 2019-06-28 at 1 29 40 PM](https://user-images.githubusercontent.com/1202812/60360091-de9e9100-99a8-11e9-81fb-5d2f48aeb3e8.png)

I'm not 100% sure it makes sense to include the top/bottom control there, but maybe? I'd love to get thoughts from another designer too. 

![Screen-Shot-2019-06-28-at-1 31 26-PM](https://user-images.githubusercontent.com/1202812/60360582-fe828480-99a9-11e9-9be8-6723d5b7594e.jpg)

Moving this control into the contextual toolbar may make it more easily adoptable for other blocks that have captions too.

melchoyce commented:
What if captions were an InnerBlock, so you could use the existing arrows to on the side of the block to use them? Too out there?

jasmussen commented:
> Not sure if there are any guides on how many icons on the toolbar is too many.

There isn't really, because the initial goal was for the block toolbar to be completely responsive, i.e. items would pop off into an overflow menu as space became a premium. In absence of that, toolbar groups can collapse, for example the alignments can collapse into a dropdown for those. 

If you have a lot of items to add to the table, consider adding them to a dropdown from the start, especially if they aren't super primary actions.

talldan commented:
> What if captions were an InnerBlock, so you could use the existing arrows to on the side of the block to move them? Too out there?

@melchoyce I think that'd pretty cool, but it's a tricky one as the html spec says the `<caption>` element can only be the first element inside a `<table>` and it should only be repositioned using the `caption-side` css property. I don't think it'd work well with the movers which are all about swapping the position of the actual elements.

jasmussen commented:
The caption position could be a style variation. I know we have stripes as a style variation, but imagine a sidebar panel called "Caption Position" with however many style-variation-like buttons showing the caption where it needs to be.

kjellr commented:
> The caption position could be a style variation. I know we have stripes as a style variation, but imagine a sidebar panel called "Caption Position" with however many style-variation-like buttons showing the caption where it needs to be.

Hmm — that's interesting. That generally makes a lot of sense. This would have to be separate from the standard block styles, since it would modify the DOM (as opposed to just being a CSS change). 

If we're doing all that though, I wonder if it's too complicated? 😄 Since there will only ever be two options for this (right?), maybe a single toolbar controls is the best option.

talldan commented:
I don't think a style variation would need to affect the DOM, the caption position is just a CSS property. Despite that, it does feel a bit like layout instead of simply being decorative.

What might be needed is a way to specify a *group* for a style variation so that separate style concepts like stripes and caption position can be selected in isolation.

Another thing that might be a bit unusual with making it a style variation is that it'd be possible for a block to have a class name like `is-style-caption-bottom` without the block even having a caption. I think that's ok though.

jasmussen commented:
> I don't think a style variation would need to affect the DOM, the caption position is just a CSS property. Despite that, it does feel a bit like layout instead of simply being decorative.

To clarify, I was suggesting it be a style variation in spirit only. I was actually thinking of copying the layout of the sidebar styles panel for a separate toggle control, basically, just using the same interface.

But the meta point is, I'm not personally convinced the caption position is a primary interaction — it definitely feels secondary to me, meaning it's okay to be sidebar only.

ellatrix commented:
Hello everyone!

Nice to see work on this. In my opinion, we should be consistent with other document _figures_ and use the `figure` and `figcaption` elements, which can be used as well according to the spec. Table happens to have its own element as well, but nothing else does. I could imagine we have more _figure_ blocks in the future that are media or display/visualise data. Consistency across all of these would be good.

Also regarding options to position the caption: we don't have this for other figure blocks. Why complicate things and add it for the table block?

And if voiceover or other assistive technology doesn't correctly associate the caption with the table, it there any way we can make it so? How does this work for images etc.?

talldan commented:
Hi @ellatrix. Thanks for the feedback. The best summary of the accessibility situation is in this comment - https://github.com/WordPress/gutenberg/pull/15554#issuecomment-500373185, and the resulting comment from Andrea -  https://github.com/WordPress/gutenberg/pull/15554#issuecomment-500402825.

`figcaption` doesn't work for tables when used with screen readers from what I could tell. It's not announced at all, and at least not in the nice way that `caption` is. I don't see an issue with using `caption`. It shouldn't cause issues with the changes at https://github.com/WordPress/gutenberg/pull/16324. A `figure` without a `figcaption` is fine.

> Also regarding options to position the caption: we don't have this for other figure blocks. Why complicate things and add it for the table block?

I think you're right, I might split it from this PR and move it to a separate PR, as it seems like the most controversial aspect.

ellatrix commented:
If only the screenreader is the problem, is there any way we can make it announce the `figcaption`? There's ways to do that right?

afercia commented:
Yes, there are techniques to do that. See example 3 here, which wraps a table within a figure element: https://www.w3.org/TR/html52/tabular-data.html#sec-techniques-for-describing-tables

See also https://www.w3.org/TR/html52/tabular-data.html#the-caption-element
> When a table element is the only content in a figure element other than the figcaption, the caption element should be omitted in favor of the figcaption.

However, there should always be a way to use an actual `<caption>` element.

talldan commented:
@afercia I tried using `figcaption` here and it wasn't announced by voiceover:
https://github.com/WordPress/gutenberg/pull/15554#issuecomment-500373185

talldan commented:
Just double-checked. There's distinct difference between the way `caption` and `figcaption` work in voiceover. 

`figcaption` is an element that can be navigated to, and it's read when it individually receives navigation, but at no other time.

`caption` is quite different in that it can't be navigated to, but is read whenever entering the table as part of the description of the table. It's implicit to the table.

There might be other instances when the table `caption` is read, hard to say as I'm not a pro user..

afercia commented:
Correct 🙂 `<caption>` is announced together with the table.

Quick tests with _Safari_ (reminder: don't test with Chrome) and VoiceOver. Navigate content using Control+Option+Right arrow (or Left arrow to go back)

caption 
https://codepen.io/afercia/full/dBaMbL

![caption](https://user-images.githubusercontent.com/1682452/60964543-e1d43e00-a313-11e9-9997-00ebc7873d4c.png)

figure with figcaption and table
https://codepen.io/afercia/full/mZvVgN

![figure with figcaption and table](https://user-images.githubusercontent.com/1682452/60964560-ea2c7900-a313-11e9-8fed-e7b4387911aa.png)

figure labelled by the table caption
https://codepen.io/afercia/full/QXYyZE

![figure labelled by the table caption](https://user-images.githubusercontent.com/1682452/60964565-edc00000-a313-11e9-9ed9-7f89a186dea6.png)

tellthemachines commented:
Regarding using `figcaption` instead of a table `caption`, would this also solve the problems you mentioned [here](https://github.com/WordPress/gutenberg/pull/15554#issuecomment-500373185) withhaving the RichText component inside the `caption`? 

If so, we might consider using the `figcaption` with an `aria-labelledby` on the table element, as proposed in [this tutorial in the WAI section of w3c](https://www.w3.org/WAI/tutorials/tables/caption-summary/#using-the-figure-element-to-mark-up-a-table-summary). 
* I checked this on Safari + VoiceOver with @afercia's codepen and it works just like a `caption` would.  
* Firefox + NVDA repeats the `figcaption` when `aria-labelledby` is used but works fine with the unlabelled `figcaption` 🤷‍♀ 

I don't have JAWS to test on though. 

Thoughts on this @afercia ? Or anyone?

afercia commented:
@tellthemachines I can test on JAWS, if necessary :)

Is the double announcement happening because reading the content sequentially:
- it first read the actual figcaption
- then reads the table, which is labelled-by the figcaption?

If so, I'd be fine with that. Sone redundancy is better than resounding silence :) Worth reminding screen readers provide tools to jump through different element types and when landing directly on a table, the figure would be "skipped". Therefore, we can't rely on the figcaption and the table needs an aria-labelledby (whether it points to a caption, figcaption, or whatever).

afercia commented:
Codepen based on the W3C tutorial example: https://codepen.io/afercia/full/WqmbPQ

With Safari and VoiceOver:
press Control+Option+Cmd+T to jump directly to the table

On Windows with NVDA and JAWS:
press T to jump directly to the table

Yes: NVDA, with both Firefox and Chrome, announces the figcaption twice even when directly jumping to the table. Also, it announces the figure as a generic "grouping".

JAWS works fine, announces just once. Same Safari + VoiceOver.

I'd say the NVDA double announcement it's a minor annoyance and seems to be a NVDA thing. Hopefully they will improve it. Will create an issue upstream.

talldan commented:
Great, thanks so much for the support on this one @afercia and @tellthemachines.

It looks like figcaption + aria-labelledby is the best option, I'll change the implementation to that.

afercia commented:
For completeness, same test jumping directly to the table:

Internet Explorer 11 and NVDA: the figcaption doesn't get announced
Internet Explorer 11 and JAWS: only the figcaption part referenced by `aria-labelledby` gets announced, while the one referenced by `aria-describedby` is ignored.

I'd be fine with this, as in the first case users have some text _before_ the table anyways.

afercia commented:
Created https://github.com/nvaccess/nvda/issues/9914 upstream.

ellatrix commented:
Thanks! Great to hear `figcaption` + `aria-labelledby` will work!

talldan commented:
> Regarding using figcaption instead of a table caption, would this also solve the problems you mentioned here withhaving the RichText component inside the caption?

This still seems to be a problem. I have to use a separate figcaption and RichText for it to work correctly.

I wondered how it was working for the image block, but in testing it's not actually working for the image block as it should be (in the editor) 😟

ellatrix commented:
@talldan What's the problem with `RichText`? What's broken in the image block? I don't follow.

talldan commented:
@ellatrix Voiceover doesn't announce the figure surrounding the image in the editor.

The cause seems to be the extra `divs` that `RichText` wraps around the `figcaption`.

Making the figcaption a direct child of the figure seems like the logical alternative:
```
<figure>
    <table> ... </table>
    <figcaption>
        <RichText />
    </figcaption>
</figure>
```
The unusual issue I had with that is the button labels for formatting options then become part of the figcaption. Voiceover will literally read "Bold Italic Link More Rich Text Controls Caption text, figure".

talldan commented on packages/block-library/src/table/save.js:
```diff
@@ -56,11 +60,18 @@ export default function save( { attributes } ) {
 			</table>
+			{ hasCaption && (
+				<RichText.Content
+					id={ captionId }
```
I'm not sure about this approach for setting an id in the `save` function.

One problem might be if a post with an existing table is edited and a new table block added, the id for that new block could match this one (as the ids will have been generated in different editor sessions).

Using the clientId of the block might be a better idea.

talldan commented on packages/block-library/src/table/edit.js:
```diff
@@ -500,15 +506,36 @@ export class TableEdit extends Component {
+						onChange={ ( value ) => setAttributes( { caption: value, captionId } ) }
+						// Deselect the selected table cell when the caption is focused.
+						unstableOnFocus={ () => this.setState( { selectedCell: null } ) }
+					/>
```
I really don't like this hack with the separate `figcaption` and `RichText` elements. I'm not sure if this was a known issue in the image block and accepted that it wouldn't work.

ellatrix replied:
Why not wrap the `<RichText>` instance in a `<figure>` element, and use a `div` as the `tagName`? One day we'll no longer have all the wrapping tags, but until that day we could do it like that?

ellatrix replied:
> One day we'll no longer have all the wrapping tags

Would love you to have a look at #17607, which will allow the table block to omit the wrapper div.

ellatrix commented:
> The cause seems to be the extra divs that RichText wraps around the figcaption.
> 
> Making the figcaption a direct child of the figure seems like the logical alternative:

Good to know. I have plans for `RichText` to be one single element in the future, without any wrapping divs. But at the movement that's a limitation. Is there no way you can link them for now?

Once I rewrite `Autocomplete`, we can get rid of the wrapper divs.

ellatrix commented:
Or it can be done as you suggest for now. :) Then RichText will be a div.

ellatrix commented on packages/block-library/src/table/save.js:
```diff
@@ -64,11 +68,18 @@ export default function save( { attributes } ) {
 	return (
 		<figure>
-			<table className={ classes }>
+			<table className={ classes } aria-labelledby={ hasCaption ? captionId : undefined }>
```
A comment here would be good I think. :)

talldan replied:
Done 👍

ellatrix commented on packages/block-library/src/table/theme.scss:
```diff
@@ -7,4 +7,10 @@
 		word-break: normal;
 	}
+
+	caption {
```
This rule is obsolete, right?

talldan replied:
Ah, yes. The selector should be `figcaption`. I've changed it over and removed some of the styling that's not needed any more.

ellatrix commented on packages/block-library/src/table/edit.js:
```diff
@@ -541,6 +544,9 @@ export class TableEdit extends Component {
 		} );
 
+		// Append the client id to the caption id to ensure no two table block captions can have clashing ids.
+		const captionId = `wp-block-table-caption-${ clientId }`;
```
Does that mean that every time a post is saved, a new ID is saved?

talldan replied:
Fixed by only creating a new id if the existing one is undefined:
https://github.com/WordPress/gutenberg/pull/15554/files#diff-3b49e2fddc9b1f6c3fbd064df1e8af10R513

I did find it quite tricky to try to use an id in this block. I think this is the first time it's been done in a core static block.

ellatrix commented on packages/block-library/src/table/deprecated.js:
```diff
@@ -8,18 +8,111 @@ import classnames from 'classnames';
 
 const deprecated = [
 	{
-		attributes: metadata.attributes,
```
Would it be possible to filter out the new attributes, instead of duplicating the whole object?

talldan replied:
I'm not a fan of that approach as it seems quite error prone for someone adding a new deprecation. Once these attributes are defined for a deprecation ideally they won't ever change.

ellatrix reviewed (dismissed):
Looks good! I'm mostly concerned about saving the clientId in the post content, which would change on every editor page load. Other than that I would suggest to not use `figure` for `RichText` and instead wrap it.

talldan commented:
> Other than that I would suggest to not use figure for RichText and instead wrap it.

I've found that it doesn't work with Voiceover, unfortunately. I've tried a bunch of things, but the only thing that seems to work reliably is a `figcaption` that only has text content within it.

Will work on the other things mentioned, thanks for reviewing 😄

talldan committed:
Reuse the captionId when already set as an attribute

talldan committed:
Change style rule to use figcaption selector

talldan committed:
Explain use of aria-labelledby

talldan committed:
Update deprecation

talldan commented:
@ellatrix This should be ready for a re-review. Thanks!

ellatrix commented:
@talldan Could you rebase this PR after #17607? It should make the rich text shadow hacks unnecessary. Then I think this PR is good to go.

talldan committed:
Use RichText for figcaption, strip out a11y hacks

talldan commented:
@ellatrix I've brought in the latest changes, and it seems to work. I need to update some tests, but it should be good to re-review soon.

The bad news is the Voiceover issues remain. The main problem is that Voiceover reads "Write caption...", the aria-label for the figcaption instead of the text content ... Not sure why.

I've decided to remove all the experimentation I did to try to get Voiceover working, and defer that to a later PR.

talldan committed:
Ensure empty classname is not output onto table element and update snapshots

talldan committed:
Update raw handling test

patrikhuber commented:
Hi! This is great. Any estimates on when we can see this in official WordPress?
Thanks!

ellatrix commented:
@talldan Looks good. There's just a failing test left?

ellatrix reviewed (approved):
Looks good. This should include a fixture for the deprecation(s) though. Why is there a deprecation is the first place?

talldan commented:
@ellatrix Thanks for the follow-up review. 

There's a very minor fix in this PR that required the deprecation. I think it's best to move that over to another PR so that it can be reviewed separately.

talldan committed:
Revert "Update raw handling test"

This reverts commit f3ae75194873ba8e72aa8980194bbe054dc75719.

talldan committed:
Revert "Ensure empty classname is not output onto table element and update snapshots"

This reverts commit 0265104ecd3741f0a9ee98e6c337250a097e0011.

talldan committed:
Update table caption fixtures and deprecations

afercia commented:
At some point there was consensus the best option was:

> It looks like figcaption + aria-labelledby is the best option

See https://github.com/WordPress/gutenberg/pull/15554#issuecomment-510379625

Also, see https://github.com/WordPress/gutenberg/pull/15554#issuecomment-500402825

> The important thing is that in the front end it's rendered as a real `<caption>` element.

Maybe I'm missing something but seems to me there's no `aria-labelledby`, nor the caption is rendered in a `<caption>` element on the front end. Is there any follow-up issue to address these points? /Cc @talldan 👋 (when you have a chance, thanks!)

paaljoachim commented:
Hey Dan @talldan 

Could something similar be used for inline alignment of image captions? https://github.com/WordPress/gutenberg/issues/12997#issuecomment-508855645

For at the moment the caption begins on the left bottom of the image. It would be great to have inline captions to choose where to have the caption. Left - center - right.

Thanks!

cirkut commented:
I want to go back to @afercia and their comment. Captions should be using the native `<caption>` element that are supported by tables. This will allow themes to directly and easily support native captions and caption placements for tables, and additionally provide the proper semantics for tables.
